File: platforms/windows/Zylix/ZylixRouter.cs
// ZylixRouter.cs - Windows Router for Zylix v0.3.0
//
// Provides WinUI 3 NavigationView integration for Zylix routing system.
// Features:
// - NavigationView/Frame integration
// - Deep link handling (Protocol activation)
// - Route parameters and query strings
// - Navigation guards
// - Back button handling

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Zylix
{
    // ========================================================================
    // Route Definition
    // ========================================================================

    /// <summary>
    /// A route parameter extracted from URL
    /// </summary>
    public record RouteParam(string Name, string Value);

    /// <summary>
    /// Parsed URL components
    /// </summary>
    public record ParsedURL(
        string Path,
        List<RouteParam> Params,
        Dictionary<string, string> Query,
        string? Fragment
    )
    {
        public string? GetParam(string name) =>
            Params.FirstOrDefault(p => p.Name == name)?.Value;

        public string? GetQuery(string key) =>
            Query.TryGetValue(key, out var value) ? value : null;
    }

    /// <summary>
    /// Route guard result
    /// </summary>
    public abstract record GuardResult
    {
        public record Allow() : GuardResult;
        public record Deny(string? Message = null) : GuardResult;
        public record Redirect(string To) : GuardResult;
    }

    /// <summary>
    /// Route metadata
    /// </summary>
    public record RouteMeta(
        string? Title = null,
        bool RequiresAuth = false,
        List<string>? Permissions = null,
        string? Icon = null,
        bool ShowInNavigation = true
    );

    /// <summary>
    /// A single route definition
    /// </summary>
    public class Route
    {
        public Guid Id { get; } = Guid.NewGuid();
        public string Path { get; set; } = "/";
        public RouteMeta Meta { get; set; } = new();
        public List<Func<RouteContext, GuardResult>> Guards { get; set; } = new();
        public List<Route> Children { get; set; } = new();
        public Type? PageType { get; set; }
    }

    /// <summary>
    /// Context passed to route handlers
    /// </summary>
    public class RouteContext
    {
        public ParsedURL Url { get; set; }
        public ZylixRouter? Router { get; set; }
        public bool IsAuthenticated { get; set; }
        public List<string> UserRoles { get; set; } = new();
        public object? UserData { get; set; }

        public RouteContext(ParsedURL url, ZylixRouter? router = null)
        {
            Url = url;
            Router = router;
        }

        public bool HasRole(string role) => UserRoles.Contains(role);
    }

    // ========================================================================
    // Navigation Event
    // ========================================================================

    public enum NavigationEvent
    {
        Push,
        Replace,
        Back,
        Forward,
        DeepLink
    }

    // ========================================================================
    // Router
    // ========================================================================

    /// <summary>
    /// Main router class for Zylix Windows
    /// </summary>
    public class ZylixRouter
    {
        private List<Route> _routes = new();
        private readonly List<string> _history = new();
        private int _historyIndex = -1;
        private readonly List<Action<NavigationEvent, string, RouteContext>> _navigationCallbacks = new();
        private Action<ParsedURL>? _notFoundHandler;
        private string _basePath = "";

        public string CurrentPath { get; private set; } = "/";
        public Route? CurrentRoute { get; private set; }
        public RouteContext? CurrentContext { get; private set; }
        public Frame? NavigationFrame { get; set; }
        public NavigationView? NavigationView { get; set; }

        public event EventHandler<RouteContext>? NavigationChanged;

        // ====================================================================
        // Configuration
        // ====================================================================

        public ZylixRouter DefineRoutes(List<Route> routes)
        {
            _routes = routes;
            return this;
        }

        public ZylixRouter SetBasePath(string path)
        {
            _basePath = path;
            return this;
        }

        public ZylixRouter OnNotFound(Action<ParsedURL> handler)
        {
            _notFoundHandler = handler;
            return this;
        }

        public ZylixRouter OnNavigate(Action<NavigationEvent, string, RouteContext> callback)
        {
            _navigationCallbacks.Add(callback);
            return this;
        }

        public IEnumerable<Route> GetNavigationRoutes() =>
            _routes.Where(r => r.Meta.ShowInNavigation);

        // ====================================================================
        // Navigation
        // ====================================================================

        public void Push(string path, object? userData = null)
        {
            Navigate(path, NavigationEvent.Push, userData: userData);
        }

        public void Replace(string path, object? userData = null)
        {
            Navigate(path, NavigationEvent.Replace, userData: userData);
        }

        public void Back()
        {
            if (!CanGoBack) return;
            _historyIndex--;
            var path = _history[_historyIndex];
            Navigate(path, NavigationEvent.Back, updateHistory: false);
        }

        public void Forward()
        {
            if (!CanGoForward) return;
            _historyIndex++;
            var path = _history[_historyIndex];
            Navigate(path, NavigationEvent.Forward, updateHistory: false);
        }

        public bool CanGoBack => _historyIndex > 0;
        public bool CanGoForward => _historyIndex < _history.Count - 1;

        // ====================================================================
        // Deep Linking
        // ====================================================================

        public void HandleDeepLink(Uri uri)
        {
            var path = string.IsNullOrEmpty(uri.AbsolutePath) ? "/" : uri.AbsolutePath;
            Navigate(path, NavigationEvent.DeepLink);
        }

        public void HandleProtocolActivation(Uri uri)
        {
            // Handle zylix:// protocol
            if (uri.Scheme.Equals("zylix", StringComparison.OrdinalIgnoreCase))
            {
                var path = "/" + uri.Host + uri.AbsolutePath;
                Navigate(path, NavigationEvent.DeepLink);
            }
        }

        // ====================================================================
        // URL Parsing
        // ====================================================================

        public ParsedURL ParseURL(string urlString)
        {
            var path = urlString;
            string? fragment = null;
            string? queryString = null;

            // Extract fragment
            var hashIndex = path.IndexOf('#');
            if (hashIndex >= 0)
            {
                fragment = path[(hashIndex + 1)..];
                path = path[..hashIndex];
            }

            // Extract query string
            var queryIndex = path.IndexOf('?');
            if (queryIndex >= 0)
            {
                queryString = path[(queryIndex + 1)..];
                path = path[..queryIndex];
            }

            // Parse query parameters
            var query = new Dictionary<string, string>();
            if (!string.IsNullOrEmpty(queryString))
            {
                foreach (var pair in queryString.Split('&'))
                {
                    var parts = pair.Split('=', 2);
                    if (parts.Length == 2)
                    {
                        query[parts[0]] = Uri.UnescapeDataString(parts[1]);
                    }
                }
            }

            return new ParsedURL(path, new List<RouteParam>(), query, fragment);
        }

        // ====================================================================
        // Route Matching
        // ====================================================================

        public (Route Route, List<RouteParam> Params)? MatchRoute(string path)
        {
            foreach (var route in _routes)
            {
                var match = MatchPattern(route.Path, path);
                if (match != null)
                {
                    return (route, match);
                }

                // Check children
                foreach (var child in route.Children)
                {
                    var fullPattern = route.Path + child.Path;
                    match = MatchPattern(fullPattern, path);
                    if (match != null)
                    {
                        return (child, match);
                    }
                }
            }
            return null;
        }

        private List<RouteParam>? MatchPattern(string pattern, string path)
        {
            var patternParts = pattern.Split('/', StringSplitOptions.RemoveEmptyEntries);
            var pathParts = path.Split('/', StringSplitOptions.RemoveEmptyEntries);

            if (patternParts.Length != pathParts.Length) return null;

            var parameters = new List<RouteParam>();

            for (int i = 0; i < patternParts.Length; i++)
            {
                var patternPart = patternParts[i];
                var pathPart = pathParts[i];

                if (patternPart.StartsWith(':'))
                {
                    var paramName = patternPart[1..];
                    parameters.Add(new RouteParam(paramName, pathPart));
                }
                else if (patternPart == "*")
                {
                    parameters.Add(new RouteParam("wildcard", pathPart));
                }
                else if (!patternPart.Equals(pathPart, StringComparison.OrdinalIgnoreCase))
                {
                    return null;
                }
            }

            return parameters;
        }

        // ====================================================================
        // Private Navigation
        // ====================================================================

        private void Navigate(
            string path,
            NavigationEvent navEvent,
            bool updateHistory = true,
            object? userData = null)
        {
            var fullPath = _basePath + path;
            var parsed = ParseURL(fullPath);

            // Match route
            var matched = MatchRoute(parsed.Path);
            if (matched == null)
            {
                _notFoundHandler?.Invoke(parsed);
                return;
            }

            var (route, parameters) = matched.Value;

            // Update parsed URL with params
            parsed = parsed with { Params = parameters };

            // Create context
            var context = new RouteContext(parsed, this)
            {
                UserData = userData
            };

            // Check guards
            foreach (var guard in route.Guards)
            {
                var result = guard(context);
                switch (result)
                {
                    case GuardResult.Allow:
                        continue;
                    case GuardResult.Deny deny:
                        System.Diagnostics.Debug.WriteLine($"[ZylixRouter] Navigation denied: {deny.Message}");
                        return;
                    case GuardResult.Redirect redirect:
                        Replace(redirect.To);
                        return;
                }
            }

            // Update history
            if (updateHistory && (navEvent == NavigationEvent.Push || navEvent == NavigationEvent.DeepLink))
            {
                // Remove forward history
                if (_historyIndex < _history.Count - 1)
                {
                    _history.RemoveRange(_historyIndex + 1, _history.Count - _historyIndex - 1);
                }
                _history.Add(path);
                _historyIndex = _history.Count - 1;
            }

            // Update state
            CurrentPath = path;
            CurrentRoute = route;
            CurrentContext = context;

            // Navigate Frame if available
            if (NavigationFrame != null && route.PageType != null)
            {
                if (navEvent == NavigationEvent.Replace && NavigationFrame.CanGoBack)
                {
                    NavigationFrame.GoBack();
                }
                NavigationFrame.Navigate(route.PageType, context);
            }

            // Notify callbacks
            foreach (var callback in _navigationCallbacks)
            {
                callback(navEvent, path, context);
            }

            NavigationChanged?.Invoke(this, context);
        }
    }

    // ========================================================================
    // Common Guards
    // ========================================================================

    public static class RouteGuards
    {
        public static GuardResult RequireAuth(RouteContext context)
        {
            if (context.IsAuthenticated)
            {
                return new GuardResult.Allow();
            }
            return new GuardResult.Redirect("/login");
        }

        public static Func<RouteContext, GuardResult> RequireRole(string role)
        {
            return context =>
            {
                if (context.HasRole(role))
                {
                    return new GuardResult.Allow();
                }
                return new GuardResult.Deny("Insufficient permissions");
            };
        }
    }

    // ========================================================================
    // NavigationView Helper
    // ========================================================================

    public static class ZylixNavigationHelper
    {
        public static void SetupNavigationView(
            NavigationView navigationView,
            Frame contentFrame,
            ZylixRouter router)
        {
            router.NavigationFrame = contentFrame;
            router.NavigationView = navigationView;

            // Add navigation items from routes
            foreach (var route in router.GetNavigationRoutes())
            {
                var item = new NavigationViewItem
                {
                    Content = route.Meta.Title ?? route.Path,
                    Tag = route.Path
                };

                if (!string.IsNullOrEmpty(route.Meta.Icon))
                {
                    item.Icon = new SymbolIcon((Symbol)Enum.Parse(typeof(Symbol), route.Meta.Icon));
                }

                navigationView.MenuItems.Add(item);
            }

            // Handle selection
            navigationView.SelectionChanged += (sender, args) =>
            {
                if (args.SelectedItem is NavigationViewItem item && item.Tag is string path)
                {
                    router.Push(path);
                }
            };

            // Handle back button
            navigationView.BackRequested += (sender, args) =>
            {
                if (router.CanGoBack)
                {
                    router.Back();
                }
            };

            // Update back button visibility
            router.NavigationChanged += (sender, context) =>
            {
                navigationView.IsBackEnabled = router.CanGoBack;
            };
        }
    }
}

File: platforms/ios/Zylix/Sources/ZylixRouter.swift
// ZylixRouter.swift - iOS Router for Zylix v0.3.0
//
// Provides SwiftUI NavigationStack integration for Zylix routing system.
// Features:
// - NavigationStack/NavigationPath integration
// - Deep link handling (Universal Links)
// - Route parameters and query strings
// - Navigation guards
// - Type-safe routing

import SwiftUI
import Combine

// MARK: - Route Definition

/// A route parameter extracted from URL
public struct RouteParam: Equatable, Hashable {
    public let name: String
    public let value: String
}

/// Parsed URL components
public struct ParsedURL: Equatable {
    public let path: String
    public let params: [RouteParam]
    public let query: [String: String]
    public let fragment: String?

    public func getParam(_ name: String) -> String? {
        params.first { $0.name == name }?.value
    }

    public func getQuery(_ key: String) -> String? {
        query[key]
    }
}

/// Route guard result
public enum GuardResult {
    case allow
    case deny(message: String?)
    case redirect(to: String)
}

/// Route metadata
public struct RouteMeta {
    public var title: String?
    public var requiresAuth: Bool = false
    public var permissions: [String] = []

    public init(title: String? = nil, requiresAuth: Bool = false, permissions: [String] = []) {
        self.title = title
        self.requiresAuth = requiresAuth
        self.permissions = permissions
    }
}

/// A single route definition
public struct Route: Identifiable {
    public let id = UUID()
    public let path: String
    public let meta: RouteMeta
    public var guards: [(RouteContext) -> GuardResult] = []
    public var children: [Route] = []

    public init(
        path: String,
        meta: RouteMeta = RouteMeta(),
        guards: [(RouteContext) -> GuardResult] = [],
        children: [Route] = []
    ) {
        self.path = path
        self.meta = meta
        self.guards = guards
        self.children = children
    }
}

/// Context passed to route handlers
public class RouteContext: ObservableObject {
    @Published public var url: ParsedURL
    @Published public var isAuthenticated: Bool = false
    @Published public var userRoles: [String] = []
    public weak var router: ZylixRouter?
    public var userData: Any?

    public init(url: ParsedURL, router: ZylixRouter? = nil) {
        self.url = url
        self.router = router
    }

    public func hasRole(_ role: String) -> Bool {
        userRoles.contains(role)
    }
}

// MARK: - Navigation Event

public enum NavigationEvent {
    case push
    case replace
    case back
    case forward
    case deepLink
}

// MARK: - Router

@MainActor
public class ZylixRouter: ObservableObject {
    @Published public var currentPath: String = "/"
    @Published public var navigationPath = NavigationPath()
    @Published public var currentContext: RouteContext?

    private var routes: [Route] = []
    private var history: [String] = []
    private var historyIndex: Int = -1
    private var navigationCallbacks: [(NavigationEvent, String, RouteContext) -> Void] = []
    private var notFoundHandler: ((ParsedURL) -> Void)?
    private var basePath: String = ""

    public init() {}

    // MARK: - Configuration

    public func defineRoutes(_ routes: [Route]) -> ZylixRouter {
        self.routes = routes
        return self
    }

    public func setBasePath(_ path: String) -> ZylixRouter {
        self.basePath = path
        return self
    }

    public func onNotFound(_ handler: @escaping (ParsedURL) -> Void) -> ZylixRouter {
        self.notFoundHandler = handler
        return self
    }

    public func onNavigate(_ callback: @escaping (NavigationEvent, String, RouteContext) -> Void) -> ZylixRouter {
        navigationCallbacks.append(callback)
        return self
    }

    // MARK: - Navigation

    public func push(_ path: String, userData: Any? = nil) {
        navigate(path, event: .push, userData: userData)
    }

    public func replace(_ path: String, userData: Any? = nil) {
        navigate(path, event: .replace, userData: userData)
    }

    public func back() {
        guard canGoBack else { return }
        historyIndex -= 1
        let path = history[historyIndex]
        navigate(path, event: .back, updateHistory: false)
    }

    public func forward() {
        guard canGoForward else { return }
        historyIndex += 1
        let path = history[historyIndex]
        navigate(path, event: .forward, updateHistory: false)
    }

    public var canGoBack: Bool {
        historyIndex > 0
    }

    public var canGoForward: Bool {
        historyIndex < history.count - 1
    }

    // MARK: - Deep Linking

    public func handleDeepLink(_ url: URL) {
        let path = url.path.isEmpty ? "/" : url.path
        navigate(path, event: .deepLink)
    }

    public func handleUniversalLink(_ url: URL) {
        // Extract path from universal link
        var path = url.path
        if !basePath.isEmpty && path.hasPrefix(basePath) {
            path = String(path.dropFirst(basePath.count))
        }
        if path.isEmpty { path = "/" }
        navigate(path, event: .deepLink)
    }

    // MARK: - URL Parsing

    public func parseURL(_ urlString: String) -> ParsedURL {
        var path = urlString
        var fragment: String? = nil
        var queryString: String? = nil

        // Extract fragment
        if let hashIndex = path.firstIndex(of: "#") {
            fragment = String(path[path.index(after: hashIndex)...])
            path = String(path[..<hashIndex])
        }

        // Extract query string
        if let queryIndex = path.firstIndex(of: "?") {
            queryString = String(path[path.index(after: queryIndex)...])
            path = String(path[..<queryIndex])
        }

        // Parse query parameters
        var query: [String: String] = [:]
        if let qs = queryString {
            for pair in qs.split(separator: "&") {
                let parts = pair.split(separator: "=", maxSplits: 1)
                if parts.count == 2 {
                    query[String(parts[0])] = String(parts[1])
                }
            }
        }

        return ParsedURL(path: path, params: [], query: query, fragment: fragment)
    }

    // MARK: - Route Matching

    public func matchRoute(_ path: String) -> (Route, [RouteParam])? {
        for route in routes {
            if let params = matchPattern(route.path, path: path) {
                return (route, params)
            }
            // Check children
            for child in route.children {
                let fullPattern = route.path + child.path
                if let params = matchPattern(fullPattern, path: path) {
                    return (child, params)
                }
            }
        }
        return nil
    }

    private func matchPattern(_ pattern: String, path: String) -> [RouteParam]? {
        let patternParts = pattern.split(separator: "/", omittingEmptySubsequences: true)
        let pathParts = path.split(separator: "/", omittingEmptySubsequences: true)

        guard patternParts.count == pathParts.count else { return nil }

        var params: [RouteParam] = []

        for (patternPart, pathPart) in zip(patternParts, pathParts) {
            if patternPart.hasPrefix(":") {
                let paramName = String(patternPart.dropFirst())
                params.append(RouteParam(name: paramName, value: String(pathPart)))
            } else if patternPart == "*" {
                params.append(RouteParam(name: "wildcard", value: String(pathPart)))
            } else if patternPart != pathPart {
                return nil
            }
        }

        return params
    }

    // MARK: - Private Navigation

    private func navigate(_ path: String, event: NavigationEvent, updateHistory: Bool = true, userData: Any? = nil) {
        let fullPath = basePath + path
        var parsed = parseURL(fullPath)

        // Match route
        guard let (route, params) = matchRoute(parsed.path) else {
            notFoundHandler?(parsed)
            return
        }

        // Update parsed URL with params
        parsed = ParsedURL(path: parsed.path, params: params, query: parsed.query, fragment: parsed.fragment)

        // Create context
        let context = RouteContext(url: parsed, router: self)
        context.userData = userData

        // Check guards
        for guard_ in route.guards {
            let result = guard_(context)
            switch result {
            case .allow:
                continue
            case .deny(let message):
                print("[ZylixRouter] Navigation denied: \(message ?? "No reason")")
                return
            case .redirect(let redirectPath):
                replace(redirectPath)
                return
            }
        }

        // Update history
        if updateHistory && (event == .push || event == .deepLink) {
            // Remove forward history
            if historyIndex < history.count - 1 {
                history.removeSubrange((historyIndex + 1)...)
            }
            history.append(path)
            historyIndex = history.count - 1
        }

        // Update state
        currentPath = path
        currentContext = context

        // Notify callbacks
        for callback in navigationCallbacks {
            callback(event, path, context)
        }
    }
}

// MARK: - Common Guards

public func requireAuth(_ context: RouteContext) -> GuardResult {
    if context.isAuthenticated {
        return .allow
    }
    return .redirect(to: "/login")
}

public func requireRole(_ role: String) -> (RouteContext) -> GuardResult {
    return { context in
        if context.hasRole(role) {
            return .allow
        }
        return .deny(message: "Insufficient permissions")
    }
}

// MARK: - SwiftUI Router View

public struct RouterView<Content: View>: View {
    @ObservedObject var router: ZylixRouter
    let content: (RouteContext?) -> Content

    public init(router: ZylixRouter, @ViewBuilder content: @escaping (RouteContext?) -> Content) {
        self.router = router
        self.content = content
    }

    public var body: some View {
        NavigationStack(path: $router.navigationPath) {
            content(router.currentContext)
        }
        .onOpenURL { url in
            router.handleDeepLink(url)
        }
    }
}

// MARK: - Router Link

public struct RouterLink<Label: View>: View {
    let path: String
    let label: () -> Label
    @EnvironmentObject var router: ZylixRouter

    public init(_ path: String, @ViewBuilder label: @escaping () -> Label) {
        self.path = path
        self.label = label
    }

    public var body: some View {
        Button(action: { router.push(path) }) {
            label()
        }
    }
}

// MARK: - Router Environment

private struct RouterKey: EnvironmentKey {
    static let defaultValue: ZylixRouter? = nil
}

public extension EnvironmentValues {
    var zylixRouter: ZylixRouter? {
        get { self[RouterKey.self] }
        set { self[RouterKey.self] = newValue }
    }
}

public extension View {
    func zylixRouter(_ router: ZylixRouter) -> some View {
        environment(\.zylixRouter, router)
            .environmentObject(router)
    }
}
